design-proposals/isolate-emulator-thread-with-smt-fix/isolate-emulator-thread-with-smt-fix.md

```diff
@@ -0,0 +1,143 @@
+apiVersion: kubevirt.io/v1
+kind: VirtualMachineInstance
+spec:
+  domain:
+    cpu:
+      dedicatedCpuPlacement: true
+      isolateEmulatorThread: true
+      sockets: 1
+      cores: 4
+      threads: 2
+```
+If the virt-launcher pod is scheduled to a node with hyperthreading enabled, then the VMI would rejected by kubelet with the following event::
+```
+SMT Alignment Error: requested 9 cpus not multiple cpus per core = 2
+```
+The following picture helps show how this is a problem
+<img src="smt-slignment-error-diagram.png">
+
+## Goals
+Fix [BZ#2228103](https://bugzilla.redhat.com/show_bug.cgi?id=2228103) on the Kubevirt scope.
+
+## Non  Goals
+Fix the actual scheduling of the virt-launcher to a node that will reject it.
+
+## Definition of Users
+This bug addresses all KubeVirt users running VMIs with isolateEmulatorThread on nodes with hyperthreading enabled - specifically DPDK and Realtime users.
+
+## User Stories
+A user should be able to run VMI with dedicatedCPUPlancement and isolateEmulatorThread enabled on a node with hyperthreading enabled, regardless of the amount of guest CPUs requested.
+
+## Repos
+- [kubevirt/kubevirt](https://github.com/kubevirt/kubevirt)
+
+# Bug fix design proposal
+The bug fix proposal introduces an alpha annotation `kubevirt.io/full-pcpus-only`.
+
+When the VMI is configured with `isolateEmulatorThread: true` and the above annotation, the virt-launcher pod will make sure to set an even amount of CPUs, meaning:
+
+User requested CPUs
+Extra housekeeping CPUs for pod
+Even (2,4,6,8,...)
++2
+odd(1,3,4,5,7,9,...)
++1 (as usual)
+
+Notes:
+1. In both scenarios the guest will get the same amount of CPUs it requested.
+2. For simplicity reasons, the housekeeping cgroup will contain only 1 CPU. However, this is open for discussion.
+
+Pros:
+* Cleaner user experience. No prior knowledge of CPUs added behind the scene needed. 
+* Matches the current crio convention, like `cpu-load-balancing.crio.io: disable`
+* Reduces the “wasted” CPUs to minimum per scenario, i.e. only for VMs that specifically add this annotation, and only when they ask for even CPUs. 
+* As this is an alpha annotation, it can be removed/ignored when and if solutions inside/outside Kubevirt such as CPU Manager properly prevent the situation from happening.
+
+Cons:
+* We’re adding yet another semi alpha knob to an already complicated set of knobs.
+
+# Fix Alternatives
+## Alternative #1 Adding `full-pcpus-only` option as a global configuration
+As opposed to the suggested solution where the user needs to add this annotation to every VMI, it is possible to make this behavior global. The assumption is that clusters run homogeneous Nodes, so this will eliminate the need to add this annotation every time.
+
+Pros:
+* On homogeneous Nodes will make the configuration of the VMI less complex.
+
+Cons:
+* Although some D/S distributions like OCP require that the cluster is homogeneous - it is not a common requirement accepted by the community.
```

Comment:
```suggestion
* The suggested annotation is redundant on nodes with hyperthreading disabled. A cluster-wide configuration may be wasteful for heterogeneous clusters, where node-level granularity is needed.
```

Reply:
It is important to say that this solution works on heterogeneous clusters. It is just wastes an extra vpcu even on nodes that do not require this waste.

Reply:
got it. I agree.